kernel/src/main/java/org/kframework/kompile/DefinitionStorage.java

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2019 K Team. All Rights Reserved.
+package org.kframework.kompile;
+
+import org.kframework.parser.concrete2kore.ParseCache;
+import org.kframework.utils.BinaryLoader;
+import org.kframework.utils.file.FileUtil;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author Denis Bogdanas
+ * Created on 09-Nov-19.
+ */
+public class DefinitionStorage {
+    private final BinaryLoader loader;
+    private final FileUtil files;
+
+    @Inject
+    public DefinitionStorage(BinaryLoader loader, FileUtil files) {
+        this.loader = loader;
+        this.files = files;
+    }
+
+    public File getCacheFile() {
+        return files.resolveKompiled("definitionAndCache.bin");
+    }
+
+    public void save(DefinitionAndCache definitionAndCache) {
+        if (!cacheParses(definitionAndCache.compiledDefinition.kompileOptions)) {
+            definitionAndCache.parseCaches.clear();
+        }
+        loader.saveOrDie(getCacheFile(), definitionAndCache);
+    }
+
+    public DefinitionAndCache load() {
```

Comment:
it seems to me like this method should be private, and you should have public getters that return either the definition or the cache, which in turn call the private load method. Right now you have both a call to loadOrDie and loadCache, and that seems like badly factored code to me.

Reply:
`DefinitionAndCache` guice object is required for `DefinitionStorage.save()`, called from 2 places. All problems you mention are just careful preservation of existing behavior. You can load the PR in Intellij and search for usages of all those methods.

Reply:
Will work on point 3.